opennav_following/src/following_server.cpp

```diff
@@ -0,0 +1,604 @@
+  auto timeout = rclcpp::Duration::from_seconds(detection_timeout_);
+  if (this->now() - detected.header.stamp > timeout) {
+    RCLCPP_WARN(this->get_logger(), "Lost detection or did not detect: timeout exceeded");
+    return false;
+  }
+
+  // Transform detected pose into fixed frame
+  if (detected.header.frame_id != fixed_frame_) {
+    try {
+      tf2_buffer_->transform(detected, detected, fixed_frame_,
+          tf2::durationFromSec(transform_tolerance_));
+    } catch (const tf2::TransformException & ex) {
+      RCLCPP_WARN(this->get_logger(), "Failed to transform detected object pose");
+      return false;
+    }
+  }
+
+  // The control law can oscillate if the orientation in the perception
+  // is not set correctly or has a lot of noise.
+  // Then, we skip the target orientation by pointing it
+  // in the same orientation than the vector from the robot to the object.
+  if (skip_orientation_) {
+    auto robot_pose = getRobotPoseInFrame(detected.header.frame_id);
+    double dx = detected.pose.position.x - robot_pose.pose.position.x;
+    double dy = detected.pose.position.y - robot_pose.pose.position.y;
+    double angle_to_target = std::atan2(dy, dx);
+    detected.pose.orientation = nav2_util::geometry_utils::orientationAroundZAxis(angle_to_target);
+  }
+
+  // Filter the detected pose
+  dynamic_pose_ = filter_->update(detected);
+  filtered_dynamic_pose_pub_->publish(dynamic_pose_);
+
+  // Return dynamic pose for debugging purposes
+  pose = dynamic_pose_;
+  return true;
+}
+
+bool FollowingServer::getFramePose(
+  const std::string & frame_id, geometry_msgs::msg::PoseStamped & pose)
+{
+  try {
+    // Get the transform from the target frame to the fixed frame
+    auto transform = tf2_buffer_->lookupTransform(
+      fixed_frame_, frame_id, tf2::TimePointZero, tf2::durationFromSec(transform_tolerance_));
+
+    // Convert transform to pose
+    pose.header.frame_id = fixed_frame_;
+    pose.header.stamp = transform.header.stamp;
+    pose.pose.position.x = transform.transform.translation.x;
+    pose.pose.position.y = transform.transform.translation.y;
+    pose.pose.position.z = transform.transform.translation.z;
+    pose.pose.orientation = transform.transform.rotation;
+  } catch (const tf2::TransformException & ex) {
+    RCLCPP_WARN(get_logger(),
+      "Failed to get transform for frame %s: %s", frame_id.c_str(), ex.what());
+    return false;
+  }
+
+  // Filter the detected pose
+  dynamic_pose_ = filter_->update(pose);
+  filtered_dynamic_pose_pub_->publish(dynamic_pose_);
+
+  // Return dynamic pose for debugging purposes
+  pose = dynamic_pose_;
+  return true;
+}
+
+geometry_msgs::msg::PoseStamped FollowingServer::getPoseAtDistance(
+  const geometry_msgs::msg::PoseStamped & pose, double distance)
+{
+  geometry_msgs::msg::PoseStamped forward_pose = pose;
+  const double yaw = tf2::getYaw(forward_pose.pose.orientation);
+  forward_pose.pose.position.x -= distance * cos(yaw);
```

Comment:
Ditto on vectors, I don't think substraction in hte fixed frame gets you what you want, depending on what quadrant the robot is located in.